### ISSUE TITLE
✨ Utility to merge UnoCSS classes/shortcuts

### ISSUE THREAD

**Issue by jd-solanki** [unocss/unocss#2748]
### Clear and concise description of the problem

Hi 👋🏻 

After consuming Anu's theme default preset which is based on UnoCSS and accepts the UnoCSS shortcuts to override or extend the component styles defined in preset shortcuts, I realized that when I combine preset's shortcuts & user's provided shortcuts they get merged like normal string.
However, when we allow overriding the specific shortcut, expected behavior is merged config will remove the default utlity and will use the user's utility.

```ts
defaultShortcuts: { btn: 'rounded px-4' }

export interface PresetOptions {
  shortcutOverrides?: Exclude<Preset['shortcuts'], undefined>
}

export function presetThemeDefault(options: PresetOptions = {}): Preset {
  return {
    // ...
	shortcuts: options.shortcutOverrides === undefined
	      ? shortcuts
	      : Array.isArray(options.shortcutOverrides)
	        ? [...options.shortcutOverrides, ...defaultShortcuts]
	        : [...defaultShortcuts, ...Object.entries(options.shortcutOverrides)],
  }
}
```

```ts
presetThemeDefault({ shortcutOverrides: { btn: 'px-3' } })
```

> Edit: with above preset code resulting shortcut will completely overriding the existing shortcut utlities and will result in `px-3` instead of `rounded px-3`

ATM resulting shortcut `btn` is `rounded px-4 px-3` but shouldn't it be `rounded px-3`?

Or more complex case can be when directions are involved:
```ts
// preset theme default
defaultShortcuts: { btn: 'rounded p-4' }

// users config
presetThemeDefault({ shortcutOverrides: { btn: 'px-3' } })
```

In this case resulting shortcut should be `rounded py-4 px-3`. As I don't have enough experience with UnoCSS directional shortcut merge feels like magic so IDK if this is even possible.

### Suggested solution

Utility to merge the shortcuts like [tailwind-merge](https://github.com/dcastil/tailwind-merge)

### Alternative

N/A

### Additional context

As this utlity will transform the generated/compile time code I guess achieving this is possible but IDK how complex it can get.

### Validations

- [X] Read the [Contributing Guidelines](https://github.com/unocss/unocss/blob/main/CONTRIBUTING.md).
- [X] Read the `README.md` of using the package.
- [X] Already used the [Interactive Docs](https://unocss.dev/interactive/).
- [X] Check that there isn't already an issue that request the same feature to avoid creating a duplicate.

**Comment by jd-solanki** [comment]
Can someone please reopen this and add no-stale? @zyyv

**Comment by wmertens** [comment]
Just chiming in to say that tailwind-merge already works for all the classes that look like tailwind classes. So you can use it together with unocss for that.

**Comment by devunt** [comment]
I think tailwind-merge does not work with unocss variant group?

**Comment by wmertens** [comment]
correct, but it does support plugins so maybe that can be expressed as a plugin?

**Comment by jd-solanki** [comment]
Hey Maintainers,

I'm working on v1 for [Anu](https://anu-vue.netlify.app/) which is built on top of UnoCSS. Again, I had to choose Tailwind or UnoCSS and still I prefer UnoCSS but I'm missing [tw-merge](https://github.com/dcastil/tailwind-merge) or [`cn`](https://www.youtube.com/watch?v=re2JFITR7TI) feature from UnoCSS and this is the reason why I'm going with Tailwind for initial development. I'm eargerly waiting for this feature to implement in Anu 🌠

For more info checkout this awesome video: https://www.youtube.com/watch?v=re2JFITR7TI

**Comment by antfu** [comment]
Just note that `tw-merge` is a runtime library that ships roughly 7.2KB minified JS to the client side - Which is against UnoCSS's philosophy. So I won't expect we have an official version of it anytime soon. But if you mean a build time solution I am surely open to it.

**Comment by jd-solanki** [comment]
As we already have [transformer](https://unocss.dev/transformers/compile-class) for compiling classes and @zyyv has developed [alias transformer](https://github.com/unpreset/unocss-transformer-alias) I guess transformer is valid candidate for this and is also build time solution.

What do you think?

**Comment by antfu** [comment]
What are you proposing exactly?

**Comment by jd-solanki** [comment]
> _I don't have in depth knowledge of how UnoCSS generates classes but here's my theory_

We can create an [transformer](https://unocss.dev/config/transformers) that reads the list of class (just like [compile class transfomer](https://unocss.dev/transformers/compile-class)) and filter out duplicate classes like tw-merge. For example, if button component definition has classes like "px-3 py-2" and on usage if I add `<Btn class="px-4" />` then resulting class attr value becomes "px-4 py-2" using uno transformer's magic.

The core thing, I want to say is transformer can read the class values and convert it according to uno engine so I guess it can also identify the duplicate classes and generate the string of unique class values.

**Comment by antfu** [comment]
A transformer can already be done in the userland, so I guess this is not blocking. But I guess it would be hard to identify what you describe as `button component definition has classes like px-3 py-2 and on usage if I add <Btn class="px-4" />` as it would be hard to tell if a class is applied to a certain component. But maybe you will find good ways :)

If you want any more capability of the transformer, feel free to send PR with the reasonings. Or what else requires on UnoCSS core to move this forward, or can it be close then?

**Comment by jd-solanki** [comment]
Thanks Anthony. After your reply I tried implementing it and realized that Uno won't able to know that button now has `px-4 py-2 rounded-lg shadow bg-green-400 text-white px-6` class as final class attribute value.

<img width="1068" alt="image" src="https://github.com/unocss/unocss/assets/47495003/943f27f7-18cc-4bab-877b-97e4bea6cce7">


I guess we can eliminate similar classes after class attrs are merged but UnoCSS runs transformer (statically I guess) before the final DOM (which is outside of the scope of Uno I guess).

**Comment by wmertens** [comment]
The way `@compiled` handles this is by having each atomic class be the hash of the attribute name followed by the hash of the value.

Then you can keep only the last classes for each attribute hash, which is a tiny utility function.

UnoCSS could do this by postprocessing, replacing the classes with such hashhash names, however this can lead to unintended replacements of course.

**Comment by magicdawn** [comment]
A little snippet to share 

```ts
import { uniq } from 'es-toolkit'
import { useMemo } from 'react'

export function getClassList(className?: string) {
  return uniq(
    (className || '')
      .split(' ')
      .map((x) => x.trim())
      .filter(Boolean),
  )
}

export function unoSimpleMerge(...classNames: Array<string | undefined>) {
  const classList = classNames.map(getClassList).flat().filter(Boolean)
  const map = new Map<string, string>()
  for (const cls of classList) {
    const lastHyphenIndex = cls.lastIndexOf('-')
    if (lastHyphenIndex === -1) {
      map.set(cls, cls)
      continue
    }
    const key = cls.slice(0, lastHyphenIndex)
    map.set(key, cls)
  }
  return Array.from(map.values()).join(' ')
}

/**
 * @note `useMemo` deps array.length can not change
 */
export function useUnoSimpleMerge(...classNames: Array<string | undefined>) {
  return useMemo(() => unoSimpleMerge(...classNames), [...classNames])
}
```

1. base on `lastIndexOf('-')`: not so correct, for example cann't merge `b-color-red` and `b-color-brand-primary`, but works for simple cases
if someone can figure out the complete utility prefix, u can make this better

2. usage
  - inside Component, use `className={useUnoSimpleMerge('mr-4px', props.className)}`
  - outside Component, use `className='mr-8px'` to overwrite `mr-4px`

**Comment by magicdawn** [comment]
and to mention after I paste this [snippet](https://github.com/unocss/unocss/issues/2748#issuecomment-2822376156) (Apr 23)
I have published this snippet to npm [unocss-merge](https://npm.im/unocss-merge) (Apr 25)

Still: only works for simple cases, use it with your own risk !

**Comment by judehunter** [comment]
Since variant groups are transformed at build time, wouldn't tailwind-merge just work? This seems to be the only issue here
